_posts/2018-11-29-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,392 @@
+
+  // Return the new modified schema
+  return transformSchema(schema, [
+    new FilterRootFields((_type, name) => {
+      return !whitelistedRootFields.includes(name)
+    }),
+    new FilterTypes(type => {
+      return !blacklistedTypes.includes(type.name)
+    }),
+    // snip
+  ])
+})
+```
+
+We can write tests for this by running [a query which returns all of the types][type-q] in a schema, and validating
+what exists:
+
+```js
+import { executableGravitySchema } from "../schema"
+import { getTypesFromSchema } from "lib/stitching/lib/getTypesFromSchema"
+
+it("Does not include blacklisted types", async () => {
+  const gravitySchema = await executableGravitySchema()
+  const gravityTypes = await getTypesFromSchema(gravitySchema)
+
+  expect(gravityTypes).not.toContain("Artist")
+  expect(gravityTypes).not.toContain("Artwork")
+})
+```
+
+This one is interesting, we _don't_ want the version of `Artist` and `Artwork` from Gravity's GraphQL
+implementation - because the hand-rolled `Artwork` and `Artist` types which lives in the source code of Metaphysics
+right now is a combination of many sources, and front-end-client specific code.
+
+If we allowed the `Artist` or `Artwork` to overwrite the existing implementations it would be a massively breaking
+change. For example, compare the Artwork type from [Gravity's GraphQL][grav-artwork] (5 fields) vs [Metaphysics'
+GraphQL][mp-artwork] (~90 fields) accidentally switching the types would cripple our front-ends.
+
+#### Merging Schemas
+
+There are two classes of schemas involved in our stitching. Local Schemas, which is our existing schema (e.g. the
+resolver live inside the current source code), and Remote Schemas (e.g. where you make an API request to run those
+resolvers). [Merging a schema][merging] has a pretty small API surface and doesn't mind which type of schemas you
+merge together.
+
+```js
+import { mergeSchemas as _mergeSchemas } from "graphql-tools"
+import { executableGravitySchema } from "lib/stitching/gravity/schema"
+import { executableConvectionSchema } from "lib/stitching/convection/schema"
+import { executableLewittSchema } from "lib/stitching/lewitt/schema"
+
+import localSchema from "../../schema"
+
+export const mergeSchemas = async () => {
+  const convectionSchema = await executableConvectionSchema()
+  const gravitySchema = await executableGravitySchema()
+
+  const mergedSchema = _mergeSchemas({
+    schemas: [gravitySchema, localSchema, convectionSchema]
+  })
+
+  return mergedSchema
+}
+```
+
+It's a pretty simple composition model, makes it real easy to do some verification tests using the same techniques
+as above.
+
+#### Stitching Schemas
+
+The next step from merging is stitching. Stitching is about taking the merged schemas and taking data from one and
+re-applying it via another API. For example, we have a consignments API (for when you want to sell a work at
+auction) and a consignment references the artwork's artist.
```

Comment:
We describe that we have a consignments api, but we never say that it's named convection. Throughout the code samples, we kind of use convection/consignments interchangeably - that is okay for us at Artsy, but if I was reading the code samples as an outsider, I'd wonder what all this "convection" stuff was.

Mentioning here that our consignments API is named convection might be plenty to solve this. I could also see changing the code samples to use variations of "consignment" exclusively, though that seems like a lot of work.

Reply:
yes

Reply:
Thanks, I've added a note about this